### PR TITLE
[VMware] Update SCSI controllers for VMs

### PR DESCRIPTION
**Pull request by nvazquez**
### Description

This PR allows updating the root disk controller type on a stopped VM via the 'rootDiskController' setting

Fixes: #5877 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### How Has This Been Tested?
- Start VM from a template with 'ide' root disk controller setting
- Stop VM
- Change the 'rootDiskController' setting to 'pvscsi'
- Start VM
- Verify the SCSI controllers on VMware Center is set to VMware Paravirtual 

- Stop VM
- Change the 'rootDiskController' setting to 'lsisas1068'
- Start VM
- Verify the SCSI controllers on VMware Center is set to LSI Logic SAS

### CODE REVIEW

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test centos7 vmware-67u3 keepEnv

**Comment by rohityadavcloud** [comment]
@nvazquez does it fix for dataDiskController VM property as well?

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java [hunk 1]
```diff
@@ -1805,6 +1805,14 @@ protected void ensureDiskControllers(VirtualMachineMO vmMo, Pair<String, String>
         int requiredNumScsiControllers = VmwareHelper.MAX_SCSI_CONTROLLER_COUNT - scsiControllerInfo.first();
         int availableBusNum = scsiControllerInfo.second() + 1; // method returned current max. bus number
 
+        if (DiskControllerType.getType(scsiDiskController) != scsiControllerInfo.third()) {
+            s_logger.debug(String.format("Change controller type from: %s to: %s", scsiControllerInfo.third().toString(),
+                    scsiDiskController));
+            vmMo.tearDownDevices(new Class<?>[]{VirtualSCSIController.class});
+            vmMo.addScsiDeviceControllers(DiskControllerType.getType(scsiDiskController));
```
Comment (sureshanaparti) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java:1812]: Hi @nvazquez What in case the root and data disk scsi controllers are different? root disk scsi controller considered as the first preference? Any doc needed.

Also, does this address the case when root disk controller is ide and data disk scsi controller is updated?

Comment (DaanHoogland) [reply plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java:1812]: > Also, does this address the case when root disk controller is ide and data disk scsi controller is updated?

and vice versa

**Comment by nvazquez** [comment]
@blueorangutan test centos7 vmware-67u3

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java [hunk 2]
```diff
@@ -1805,6 +1805,14 @@ protected void ensureDiskControllers(VirtualMachineMO vmMo, Pair<String, String>
         int requiredNumScsiControllers = VmwareHelper.MAX_SCSI_CONTROLLER_COUNT - scsiControllerInfo.first();
         int availableBusNum = scsiControllerInfo.second() + 1; // method returned current max. bus number
 
+        if (DiskControllerType.getType(scsiDiskController) != scsiControllerInfo.third()) {
```
[the remark this answers is not in the document]
Comment (nvazquez) [reply plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java:1808]: In most of the cases yes - the only case in which `third()` is the type of the data disk controller is when the root disk controller is IDE

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java (hunk 1 above)
Comment (nvazquez) [reply plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java:1812]: Those cases are handled by the method `HypervisorHostHelper.getScsiController()` which sets the `scsiDiskController` value depending on the root and data disk controllers and the recommended controller.

**Comment by nvazquez** [comment]
@rohityadavcloud yes, sorry for the delay. Was investigating some issues that were environmental. The data disk controller is set as the SCSI controller when the root disk controller is set to 'ide'

**Comment by nvazquez** [comment]
@sureshanaparti @DaanHoogland @shwstppr this is ready for testing/review

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/VirtualMachineMO.java [hunk 3]
```diff
@@ -2293,6 +2293,43 @@ public int getPvScsiDeviceControllerKey() throws Exception {
...
+                return new ParaVirtualSCSIController();
+            case lsilogic:
+                return new VirtualLsiLogicController();
+            case lsisas1068:
+                return new VirtualLsiLogicSASController();
+            case buslogic:
+                return new VirtualBusLogicController();
+            default:
+                return new VirtualLsiLogicController();
+        }
```
Comment (sureshanaparti) [inline vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/VirtualMachineMO.java:2308]:
```
switch (type) {
             case pvscsi:
                 return new ParaVirtualSCSIController();
             case lsisas1068:
                 return new VirtualLsiLogicSASController();
             case buslogic:
                 return new VirtualBusLogicController();
             case lsilogic:
             default:
                 return new VirtualLsiLogicController();
}
```

**Comment by sureshanaparti** [comment]
@nvazquez Performed the following test manually. Please check tests not working as expected (marked in bold).

- Deployed VM (with default CentOS template) with 3 GB DATA disk [Passed, Works as expected]
ROOT Disk => CloudStack: rootDiskController 'ide', VMware: IDE Controller
DATA Disk => CloudStack: dataDiskController 'osdefault', VMware: LSI Logic Parallel SCSI controller

**Updated DATA disk controllers:**

- Stop VM, Update setting dataDiskController to 'pvscsi', and Start VM [Passed, Works as expected]
ROOT Disk => CloudStack: rootDiskController 'ide', VMware: IDE Controller
DATA Disk => CloudStack: dataDiskController 'pvscsi',VMware: Paravirtual SCSI controller

- Stop VM, Update setting dataDiskController to 'buslogic', and Start VM [Passed, Works as expected - May be the error message 'Unable to start a VM due to insufficient capacity' has to be improved to indicate the actual error]
Failed to Start, Error in VMware: The BusLogic SCSI adapter is not supported for 64-bit guests. See the documentation for the appropriate type of SCSI adapter to use with 64-bit guests.
	
- **Stop VM, Update setting dataDiskController to 'ide', and Start VM [Failed, Not working as expected]**
Failed to Start, Error in VMware: The BusLogic SCSI adapter is not supported for 64-bit guests. See the documentation for the appropriate type of SCSI adapter to use with 64-bit guests.

- Stop VM, Update setting dataDiskController to 'lsisas1068', and Start VM [Passed, Works as expected]
ROOT Disk => CloudStack: rootDiskController 'ide', VMware: IDE Controller
DATA Disk => CloudStack: dataDiskController 'lsisas1068', VMware: LSI Logic SAS SCSI controller

- Stop VM, Update setting dataDiskController to 'lsilogic', and Start VM [Passed, Works as expected]
ROOT Disk =>  CloudStack: rootDiskController 'ide', VMware: IDE Controller
DATA Disk => CloudStack: dataDiskController 'lsilogic', VMware: LSI Logic Parallel SCSI controller

- Stop VM, Update setting dataDiskController to 'scsi', and Start VM [Passed, Works as expected]
ROOT Disk =>  CloudStack: rootDiskController 'ide', VMware: IDE Controller
DATA Disk => CloudStack: dataDiskController 'scsi', VMware: LSI Logic Parallel SCSI controller

**Updated both ROOT and DATA disk controllers:**

- Stop VM, Update setting rootDiskController to 'pvscsi' and dataDiskController to 'scsi', and Start VM [Passed, Works as expected]
ROOT Disk =>  VMware: Paravirtual SCSI controller
DATA Disk => VMware: Paravirtual SCSI controller

- Stop VM, Update setting rootDiskController to 'lsisas1068' and dataDiskController to 'pvscsi', and Start VM [Passed, Works as expected]
ROOT Disk =>  VMware: LSI Logic SAS SCSI controller
DATA Disk =>  VMware: LSI Logic SAS SCSI controller

- Stop VM, Update setting rootDiskController to 'lsisas1068' and dataDiskController to 'ide', and Start VM [Passed, Works as expected]
ROOT Disk =>  VMware: LSI Logic SAS SCSI controller
DATA Disk => VMware: IDE Controller

- Stop VM, Update setting rootDiskController to 'pvscsi' and dataDiskController to 'buslogic', and Start VM [Passed, Works as expected]
ROOT Disk =>  VMware: Paravirtual SCSI controller
DATA Disk =>  VMware: Paravirtual SCSI controller

- Stop VM, Update setting rootDiskController to 'buslogic' and dataDiskController to 'lsilogic', and Start VM [Passed, Works as expected - May be the error message has to be improved to indicate the actual error]
Failed to Start, Error in VMware: The BusLogic SCSI adapter is not supported for 64-bit guests. See the documentation for the appropriate type of SCSI adapter to use with 64-bit guests.

- Stop VM, Update setting rootDiskController to 'ide' and dataDiskController to 'lsilogic', and Start VM [Passed, Works as expected]
ROOT Disk =>  VMware: IDE controller
DATA Disk => VMware: LSI Logic Parallel SCSI controller

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/VirtualMachineMO.java (hunk 3 above)
Comment (sureshanaparti) [reply vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/VirtualMachineMO.java:2308]: @nvazquez check and address if this is ok.

Comment (nvazquez) [reply vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/VirtualMachineMO.java:2307]: Done, thanks

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
Thanks for your testing @sureshanaparti - I have examined the remaining specific issue and did not find an easy fix, do you agree on proceeding with this PR and open a new issue for the corner case to fix in 4.17?

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test matrix

**Comment by sureshanaparti** [comment]
> Thanks for your testing @sureshanaparti - I have examined the remaining specific issue and did not find an easy fix, do you agree on proceeding with this PR and open a new issue for the corner case to fix in 4.17?

agree @nvazquez

**Review by sureshanaparti — APPROVED** [review]

**Review by DaanHoogland — APPROVED** [review]
cltgm

**Comment by DaanHoogland** [comment]
as per the test report by @sureshanaparti and discussion about it with @nvazquez, I'm merging this. @nvazquez, can you comment here with a 4.17 issue number for future reference?
